src/kernels/common/baseJupyterSession.ts

```diff
@@ -335,14 +343,33 @@ export abstract class BaseJupyterSession implements IBaseKernelConnectionSession
                     }
                 };
                 session.kernel.statusChanged?.connect(handler);
+                disposables.push(
+                    new Disposable(() => swallowExceptions(() => session.kernel?.statusChanged?.disconnect(handler)))
+                );
                 if (session.kernel.status == 'idle') {
                     deferred.resolve(session.kernel.status);
                 }
-                const result = await Promise.race([deferred.promise, sleep(timeout)]);
-                session.kernel.statusChanged?.disconnect(handler);
+                // Check for possibility that kernel has died.
+                const sessionDisposed = createDeferred<unknown>();
+                session.disposed.connect(sessionDisposed.resolve, sessionDisposed);
+                disposables.push(
+                    new Disposable(() =>
+                        swallowExceptions(() => session.disposed.disconnect(sessionDisposed.resolve, sessionDisposed))
+                    )
+                );
+                const sleepPromise = sleep(timeout);
+                sessionDisposed.promise.ignoreErrors();
+                sleepPromise.ignoreErrors();
+                deferred.promise.ignoreErrors();
+                const result = await Promise.race([deferred.promise, sleepPromise, sessionDisposed.promise]);
```

Comment:
`deferred` could be named better - `kernelConnected`?

Reply:
Changed to `kernelStatus` as its a promise with the value of status